src/VisualStudio/Core/Def/Implementation/SymbolSearch/RoslynSymbolSource.cs

```diff
@@ -0,0 +1,73 @@
+        public string UniqueId => "Roslyn symbol source";
+
+        public ImageId DisplayIcon => throw new NotImplementedException();
+
+        public string DescriptionText => "Symbols provided by the local language service";
+
+        public string DisplayName => "Local symbol source";
+
+        public async Task<SymbolSearchStatus> FindSymbolsAsync(string navigationKind, VisualStudio.Language.Intellisense.SymbolSearch.Location sourceLocation, IStreamingSymbolSearchSink sink, CancellationToken token)
+        {
+            var snapshot = sourceLocation.PersistentSpan.Document.TextBuffer.CurrentSnapshot;
+            var roslynDocument = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var symbolSearchContext = new SymbolSearchContext(this, sink);
+            switch (navigationKind)
+            {
+                case PredefinedSymbolNavigationKinds.Definition:
+                    {
+                        var goToDefinitionService = roslynDocument.GetLanguageService<IGoToDefinitionService>();
+                        goToDefinitionService.TryGoToDefinition(roslynDocument, sourceLocation.PersistentSpan.Span.GetStartPoint(snapshot).Position, token);
+                        return SymbolSearchStatus.Completed;
+                    }
+                case PredefinedSymbolNavigationKinds.Implementation:
+                    {
+                        var findUsagesService = roslynDocument.GetLanguageService<IFindUsagesService>();
+                        var context = new SimpleFindUsagesContext(token);
+                        await findUsagesService.FindImplementationsAsync(roslynDocument, sourceLocation.PersistentSpan.Span.GetStartPoint(snapshot).Position, symbolSearchContext)
+                            .ConfigureAwait(false);
+                        return SymbolSearchStatus.Completed;
+                    }
+                case PredefinedSymbolNavigationKinds.Reference:
+                    {
+                        var findUsagesService = roslynDocument.GetLanguageService<IFindUsagesService>();
+                        var context = new SimpleFindUsagesContext(token);
+                        await findUsagesService.FindReferencesAsync(roslynDocument, sourceLocation.PersistentSpan.Span.GetStartPoint(snapshot).Position, symbolSearchContext)
+                            .ConfigureAwait(false);
+                        return SymbolSearchStatus.Completed;
+                    }
+                case PredefinedSymbolNavigationKinds.Base:
+                    {
+                        var goToBaseService = roslynDocument.GetLanguageService<IGoToBaseService>();
+                        var context = new SimpleFindUsagesContext(token);
+                        await goToBaseService.FindBasesAsync(roslynDocument, sourceLocation.PersistentSpan.Span.GetStartPoint(snapshot).Position, symbolSearchContext)
+                            .ConfigureAwait(false);
+                        return SymbolSearchStatus.Completed;
```

Comment:
Those cases looks so similar. Maybe some refactoring can be applied for all/some of them.